### DISCUSSION TITLE
ReactiveMongoCollection decrement array-objects

### DISCUSSION THREAD

**Discussion by nextNluwi** [quarkusio/quarkus#24625] [Q&A] [answered]
Hello everyone,

i am currently having problems with my badge implementation using the ReactiveMongoCollection.

I have the following database collection:
```
{
    "_id": "xxx",
    "badges": [
    {
        "count": 3,
        "categorieId": "menu_ma_news"
    }, 
    {
        "count": 1,
        "categorieId": "menu_ma_news",
        "subCategorieId": "1"
    }, 
    {
        "count": 2,
        "categorieId": "menu_ma_news",
        "subCategorieId": "2"
    }],
    "globalBadges": 3
}
```

I would now like to delete all badges which have categorieId = x and subCategorieId = y and at the same time get the sum of the counts to subtract them from the globalBadges and the array-object where categorieId = x.

I have found many examples for mongodb itself, but currently do not know how to implement them in java with my ReactiveMongoCollection.

For example something like this which returns only the matching array-objects:
```
db.test.aggregate([
    // Get just the docs that contain a shapes element where color is 'red'
    {$match: {'shapes.color': 'red'}},
    {$project: {
        shapes: {$filter: {
            input: '$shapes',
            as: 'shape',
            cond: {$eq: ['$$shape.color', 'red']}
        }},
        _id: 0
    }}
])
```
So i can iterate through the list, decrement them and then delete them...

Right now i have something like this:
```
public Uni<Void> deleteBadge(String categorieId, String subCategorieId)
  {
    return getCollection()
      .updateOne(
        eq("_id", "xxx"),
        Updates.pullByFilter(
          Filters.eq("badges",
            and(eq("categorieId", categorieId),
              and(eq("subCategorieId", subCategorieId)))
          )))
      .onItem().ignore().andContinueWithNull();
  }
```

Thanks in advance!

Luk

[a comment by an automated account is not shown]

**Comment by evanchooly** [comment]
Reactive APIs aside, I'm not sure mongo's query/update facilities support what you're trying to do.  You could maybe use an aggregation and merge it back to the original collection, though.

**Comment by nextNluwi** [comment]
This is exactly where I get stuck. I don't know how to implement the aggregation i have already given above in my java.

**Reply by loicmathieu** [reply]
You can write an aggregation using the Mongo Java Driver: https://www.mongodb.com/docs/drivers/java/sync/v4.5/fundamentals/aggregation

The reactive driver based on Mutiny allow the same but wrapped with Mutiny operators.

**Comment by nextNluwi** [comment, marked as answer]
Got it! Thank you guys!
```
return getCollection()
      .aggregate(
        Arrays.asList(
          new Document("$match",
            new Document("badges.categorieId", categorieId)
              .append("badges.subcategorieId", subCategorieId)
              .append("_id", jwt.getSubject())),
          new Document("$project",
            new Document("badges",
              new Document("$filter",
                new Document("input", "$badges")
                  .append("as", "badges")
                  .append("cond",
                    new Document("$eq", Arrays.asList("$$badges.subCategorieId", subCategorieId)))))))
      )
```
